Matěj z

TheMrGhostman opened the pull request:
GANomaly now can run on GPU (successfully tested).

Only thing that was not fully tested is appling results into `GenerativeAD.experiment(result..., data, savepath; save_entries...) `

TheMrGhostman committed:
integration into framework

adding training loop for ganomaly

TheMrGhostman committed:
Merge branch 'Matěj-Z' of https://github.com/aicenter/GenerativeAD.jl into Matěj-Z

TheMrGhostman committed:
train on gpu added

TheMrGhostman committed:
Update GANomaly.jl

TheMrGhostman committed:
fixing computations of anomaly scores

TheMrGhostman committed:
update

TheMrGhostman committed:
update

TheMrGhostman committed:
fix

TheMrGhostman committed:
fix

TheMrGhostman committed:
update

GANomaly can now run on GPU

TheMrGhostman committed:
forgotten scaling to interval

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -22,10 +23,28 @@ end
+end
+
+
+function check_params(savepath, parameters, data)
```
you can use GenerativeAD.check_params

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -42,42 +61,71 @@ Note:
-    end
     # prepare batches & loaders
-    train_loader = MLDataPattern.RandomBatches(X_train, parameters.batch_size, parameters.iters)
+    train_loader = Flux.Data.DataLoader(data |> gpu, batchsize=parameters.batch_size, shuffle=true)
```
Does moving all the data to gpu at once work? In my experience this leads to using up too much of the card memory (or running out of memory completely). I usually keep the data on cpu memory and only send the current batch to gpu.

vitskvara replied:
You could handle that on the level of your custom fit! function.

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -42,42 +61,71 @@ Note:
+while try_counter < max_tries
+	parameters = sample_params()
+
+    # computing additional parameters
```
All this should be done by overloading the GenerativAD.edit_params function

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -42,42 +61,71 @@ Note:
+        data = GenerativeAD.preprocess_images(data, parameters)
+        #(X_train,_), (X_val, y_val), (X_test, y_test) = data
+
+		training_info, results = fit(data[1][1], parameters)
```
Please fix the indentation to the same format.

vitskvara commented on src/models/GANomaly.jl:
```diff
@@ -266,34 +293,64 @@ function StatsBase.fit!(generator::Generator, discriminator::Discriminator, opt,
+                                        (:generator_loss, loss1[1]),
+                                        (:discriminator_loss, loss2)
+                                        ])
+            #případně přidat kontrolu velikosti chyby s restartem discriminatoru
```
Do you not use early stopping?

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -1,5 +1,6 @@
 using ArgParse
 using GenerativeAD
+using GenerativeAD.Models: anomaly_score
 using DrWatson
```
Move `using DrWatson` and `@quickactivate` to the first 2 lines of the script, that way it will always trigger the version of code that you run from,

vitskvara replied:
which is useful for testing on RCI cluster in your home folder.

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -42,42 +61,71 @@ Note:
+
+try_counter = 0
+max_tries = 2
+while try_counter < max_tries
```
Check `experiments_tabular/knn.jl`, we have changed this loop so that the loop over seeds is inside. However, for image datasets, I think it will be more important to do the loop over classes properly and only run over one seed - otherwise its going to be too computationally heavy.

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -42,42 +61,71 @@ Note:
+
+savepath = datadir("experiments/images/$(modelname)/$(dataset)/seed=$(seed)")
 
+data = GenerativeAD.load_data(dataset, seed=seed)
```
Loop over classes here, i.e. use `GenerativeAD.load_data(dataset, seed=seed, anomaly_class_ind=i)`

TheMrGhostman committed:
update

- fixing indentations (hope so)
- another input parameter -> anomaly class ind(ex)
- data goes to GPU right before forward pass
-

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -27,96 +31,87 @@ modelname = "Conv-GANomaly"
 #_________________________________________________________________________________________________
 
-savepath = datadir("experiments/images/$(modelname)/$(dataset)/seed=$(seed)")
+savepath = datadir("experiments/images/$(modelname)_ac=$(anomaly_class_ind)/$(dataset)/seed=$(seed)")
```
Append the class index to `$dataset`, not to the model name.

TheMrGhostman committed:
adding bash script

-- not tested

vitskvara commented:
Looks good

TheMrGhostman committed:
SkipGANomaly

- first interation of SkipGANomaly model

TheMrGhostman committed:
adding early stopping after epochs

- TODO: return multiple anomaly scores for SkipGAN, proper testing

TheMrGhostman committed:
update

TheMrGhostman committed:
update

TheMrGhostman committed:
fixes minor bugs

TheMrGhostman committed:
fixing minor bugs

TheMrGhostman committed:
updating early stopping and fixing bugs

- GANomaly was updated and fully tested except passing results to GenerativeAD.experimintes(...)

TheMrGhostman committed:
finishing SkipGANomaly

- SkipGANomaly was updated and fully tested except passing results to GenerativeAD.experimintes(...)
- multiple anomaly score will be added

TheMrGhostman committed:
multiple anomaly scores added

TheMrGhostman committed:
bash script added

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -10,74 +11,118 @@ using MLDataPattern
+			[0:3...],
+			[0.0001:0.0001:0.001..., 0.002:0.001:0.01...],
+            [2^x for x=2:8],
+			[10000],
```
No more epochs?

TheMrGhostman replied:
yes. one epoch was too much for GANomaly (it hits minimum val_loss within 300 iterations but epoch is about 600) . So i had to go back to iterations.

vitskvara replied:
So des that mean you never go through the whole dataset? Is that not a little limiting? Maybe for non-MNIST problems you will have to do epochs.

TheMrGhostman replied:
that is true. Reconstruction capabilities of this autoencoder with those multiple loss functions are pretty good. For other datasets we can simply set more possible iterations. But author of original paper stated that  he used 10, 15 and 25 epochs for non-MNIST datasets (25 epochs = +/- 15 000 iterations)

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -10,74 +11,118 @@ using MLDataPattern
 
-    return training_info, predict(X_val)
+	return training_info, [(x -> GenerativeAD.Models.anomaly_score(generator|>cpu, x; dims=3), parameters)]
+	# not sure if I should return generator and disciriminator in GPU
```
This depends on whether you continue to compute the scores on gpu (which should be faster), but returning the model on cpu is safer, it can always be ported back.

TheMrGhostman replied:
I thought so. I wanted to avoid unnecessary complications.

vitskvara replied:
Actually when I look at it again it cannot be ported back to gpu when its a part of the anonymous function... maybe leaving it on gpu is better, but you have to take care of having the data on gpu in evaluation as well.

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -10,74 +11,118 @@ using MLDataPattern
+max_tries = 10*max_seed
+
+
+while try_counter < max_tries
```
This is just a detail, but the indentation format is still a bit off.

TheMrGhostman replied:
I know that sometimes I have two empty lines instead of one. 
However i am not sure why this happens 
`
+        		data = GenerativeAD.Models.preprocess_images(data, parameters, denominator=32)
+        		#(X_train,_), (X_val, y_val), (X_test, y_test) = data
+                training_info, results = fit(data, parameters)
`
because in my editor it is correct.

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -10,74 +11,118 @@ using MLDataPattern
+        		# now loop over all anomaly score funs
+        		for result in results
+        			GenerativeAD.experiment(result..., data, savepath; save_entries...)
+        		end
```
You are missing the condition for ending the while loop, see `knn.jl`

TheMrGhostman replied:
I missed that ... i added `global try_counter = max_tries + 1`

vitskvara commented on scripts/experiments_images/GANomaly_run.sh:
```diff
@@ -0,0 +1,14 @@
+#SBATCH --time=24:00:00
+#SBATCH --nodes=1 --ntasks-per-node=2 --cpus-per-task=1
+#SBATCH --gres=gpu:1
+#SBATCH --mem=20G
```
Is this enough memory?

TheMrGhostman replied:
I think it is, but i can raise it.

vitskvara commented on scripts/experiments_images/GANomaly_run.sh:
```diff
@@ -0,0 +1,14 @@
+ANOMALY_CLASSES=$3
+
+module load Julia/1.4.1-linux-x86_64
+module load Python/3.8.2-GCCcore-9.3.0
```
Do you need Python? Are you using PyCall?

TheMrGhostman replied:
good point. I forgot remove it.

vitskvara commented on scripts/experiments_images/SkipGANomaly.jl:
```diff
@@ -0,0 +1,127 @@
+
+	training_info = (
+		fit_t = fit_t,
+		model = (model |> cpu),
```
Save model only once. Probably in the fit function in case of a successful fit. That means that the fit function also needs to know the savepath. The filename should consist of the parameters used in the construction of the model - it should not contain the parameters of the anomaly scoring functions. Test whether this design does not cause trouble somewhere else, e.g. in the check params function.

vitskvara commented on scripts/experiments_images/SkipGANomaly.jl:
```diff
@@ -0,0 +1,127 @@
+        		# now loop over all anomaly score funs
+        		for result in results
+        			GenerativeAD.experiment(result..., data, savepath; save_entries...)
+        		end
```
Again, missing break of the while loop.

TheMrGhostman replied:
fixed

vitskvara commented on scripts/experiments_images/SkipGANomaly_run.sh:
```diff
@@ -0,0 +1,14 @@
+ANOMALY_CLASSES=$3
+
+module load Julia/1.4.1-linux-x86_64
+module load Python/3.8.2-GCCcore-9.3.0
```
Same points as in the case of the ganomaly script

TheMrGhostman replied:
fixed

vitskvara commented on src/models/Models.jl:
```diff
@@ -10,8 +10,11 @@ using ImageTransformations
 using MLDataPattern
 using ProgressMeter: Progress, next!
 
+include("utils/utils.jl")
```
Does this need to be in a subdirectory? I think it only makes sense if we add more similar files there.

TheMrGhostman replied:
It is not needed but i did it because Honza made this directory in his branch/pull request. So i thought i can just use it.

vitskvara replied:
yeah then lets keep it

vitskvara commented on src/models/SkipGANomaly.jl:
```diff
@@ -0,0 +1,315 @@
+									(:discriminator_loss, loss2)
+									])
+		#TODO optionaly add discriminator restrart if its loss drops under 1e-5
+		if mod(iters, params.check_every) == 0
```
Inconsistent indentation, a bit hard to read. How does the early stopping work here?

TheMrGhostman replied:
Here it is bit tricky(resp messy) but while i checking checking early stopping criterion `anomality = (params.lambda*tot_val_loss_rl + (1-params.lambda)*tot_val_loss_ll)` 
i am also computing generator and discriminator loss on validation dataset with normal observations.
I thought it can be useful to trace it as well.

vitskvara replied:
I see, this makes sense.

TheMrGhostman committed:
update

TheMrGhostman committed:
indentation update

TheMrGhostman committed:
save model separately

TheMrGhostman committed:
n_params

vitskvara commented on src/models/utils/utils.jl:
```diff
@@ -0,0 +1,92 @@
+	return (X_train, y_train), (X_val, y_val), (X_test, y_test)
+end
+
+function GANomalyHistory()
```
Consider using ValueHistories.jl for this, they basically do exactly the same thing.

TheMrGhostman replied:
i will take a look on it

TheMrGhostman committed:
fix

vitskvara commented on scripts/experiments_images/GANomaly.jl:
```diff
@@ -10,74 +11,118 @@ using MLDataPattern
+
+	for seed in 1:max_seed
+		for i in 1:anomaly_classes
+			savepath = datadir("experiments/images/$(modelname)/$(dataset)_ac=$(i)/seed=$(seed)")
```
Change to `"experiments/images/$(modelname)/$(dataset)/ac=$(i)/seed=$(seed)"`

TheMrGhostman replied:
changed

TheMrGhostman committed:
update

vitskvara commented on scripts/experiments_images/SkipGANomaly.jl:
```diff
@@ -0,0 +1,129 @@
+		fit_t = fit_t,
+		model = (info[2] |> cpu),
+		history = info[1], # losses through time
+        n_parameters = info[3] # number of parameters
```
please rename this to npars

TheMrGhostman replied:
renamed

vitskvara commented on scripts/experiments_images/SkipGANomaly.jl:
```diff
@@ -0,0 +1,129 @@
+		global info, fit_t, _, _, _ = @timed fit!(model |>gpu , data, parameters)
+	catch e
+		println("Error caught.")
+		return (fit_t = NaN,), []
```
set the rest of the fields as well, e.g. `model = nothing, ...`

TheMrGhostman replied:
done

vitskvara commented on scripts/experiments_images/SkipGANomaly.jl:
```diff
@@ -0,0 +1,129 @@
+				#(X_train,_), (X_val, y_val), (X_test, y_test) = data
+				training_info, results = fit(data, parameters)
+				# saving model separately
+				tagsave(joinpath(savepath, savename("model", parameters, "bson")), Dict("model"=>training_info.model), safe = true)
```
Check if model was trained at all, e.g. by `if training_info.model != nothing`

TheMrGhostman replied:
only model saving or i should incoude the rest as well?

TheMrGhostman committed:
update

TheMrGhostman committed:
update training_info

TheMrGhostman committed:
added computing of anomaly scores on gpu

TheMrGhostman committed:
little fix

vitskvara commented:
Seems ok.

TheMrGhostman committed:
fix

vitskvara committed:
Merge branch 'master' into Matěj-Z
